### ISSUE TITLE
Huge efficiency improvement in general graph constructor

### ISSUE THREAD

**Issue by mrtzh** [JuliaLang/Graphs.jl#160]
While building a graph from a large edge list (social network data) I noticed that the general method to construct a graph (line 38 in graph.jl) is extremely slow on large graphs. Below I will explain why and how to fix it. The resulting method is faster even for tiny graphs and already gives a factor 100 improvement in running time of graphs as small as 100k vertices. For my data set it reduced the construction time from several hours to a few minutes.

Currently the graph constructor 
```julia
function graph{V,E}(vs::Vector{V}, es::Vector{E}; is_directed::Bool=true)
```
 calls `add_edge` for each edge in `es`. Unfortunately, `add_edge` contains these lines of code:

```julia
    ui = vertex_index(u, g)::Int
    vi = vertex_index(v, g)::Int
```
The function `vertex_index` simply uses `findfirst` to find the index of `u` in the `g.vertices`. As the vertex lists gets longer, this operation becomes too expensive.

A much better solution is to use a hash table for vertex indices. Below I show a better constructor:

```julia
function graphfast{V,E}(vs::Vector{V}, es::Vector{E}; is_directed::Bool=true)
  n = length(vs)
  
  # use hash table to access vertex indices
  indices = (V => Int)[]

  for i in 1:length(vs)
    indices[vs[i]]=i
  end
  
  g = Graph{V,E}(is_directed, vs, E[], multivecs(E, n), multivecs(E, n))
  
  for e in es
    u = source(e,g)
    v = target(e,g)
    ui = indices[u]
    vi = indices[v]
    
    push!(g.edges, e)
    push!(g.finclist[ui], e)
    push!(g.binclist[vi], e)

    if !g.is_directed
        rev_e = revedge(e)
        push!(g.finclist[vi], rev_e)
        push!(g.binclist[ui], rev_e)
    end
    e
  end
  return g
end
```
As you can see the main change is just a hash table for vertex indices.
Here's a little test to show how effective this change is:
```julia
function random_graph(n,m)
  vs = rand(1:10*n,n)
  es = [ Edge(i,vs[rand(1:n,1)[1]],vs[rand(1:n,1)[1]]) for i=1:m ]
  t1 = (@elapsed graph(vs,es))
  t2 = (@elapsed graphfast(vs,es))
  return (t1,t2)
end
```
Here are some numbers (from my old macbook air):
```julia
ts = [random_graph(100*i,100*i) for i=[1,10,100,1000]
```
(0.000271988,6.3622e-5)
(0.004185826,0.000618071)
(0.32407082,0.006672994)
(24.353708391,0.179466982)

So, even for just 100,000 vertices, the last line is more than a factor 100 smaller for the faster constructor. On my data set it reduced the construction time from 5 hours to a few minutes.

**Question:** I'm happy to put in a pull request for the code. It's fully compatible with the old constructor and can be replaced as is. If people want to be more conservative, we could add this in as a new method with a different name. Since I'm new here, I'd like to ask how you'd like me to go about this. Any thoughts?

**Comment by mlubin** [comment]
I think the approach consistent with the design of Graphs would be to introduce a new type of graph which implements ``vertex_index(u,g)`` by using a hash table. This way ``vertex_index`` would be fast when used inside of algorithms as well. Have you looked at using ``KeyVertex``? In this case ``vertex_index`` is implemented by a simple O(1) lookup.

**Comment by mrtzh** [comment]
Yes. I looked into `KeyVertex` first. It's just a type that stores the index with the vertex. It doesn't solve the problem of getting that index the first time around when you're processing a list of edges that isn't already made of `Edge{KeyVertex}` edges. For example, whenever you're reading edges from a file or a stream, they're typically pairs of labels without the indices. So you need to find those indices quickly, when you see a new edge.

**Comment by mlubin** [comment]
What about a constructor that generates the ``KeyVertex`` types once all the vertices are indexed?
It seems like a waste to generate this vertex-index map when constructing the graph and then just discard it.

**Comment by mrtzh** [comment]
I completely agree with you, it's probably a good idea to keep the dictionary with the graph data structure as opposed to throwing it away. 

You can build your vertices of type `KeyVertex` quickly, that's no problem. But let's say the way you're getting your edges is just a list of pairs of integer labels. Whenever you see such an edge, it doesn't help that you already have list of `KeyVertex` vertices sitting around. That list doesn't help you look up the index corresponding to an integer label quickly.

**Comment by mlubin** [comment]
If I've learned something about graphs it's that every possible way to represent them will be useful for *someone* :). So if the dictionary version makes more sense for you than ``KeyVertex``, go for it. Actually I'm not too sure why we wouldn't want to store a vertex index dictionary in the default graph type, unless we're trying to support vertex objects that aren't hashable.

**Comment by mrtzh** [comment]
I agree that a clean fix would be to put a dictionary in the default graph type.

However, I feel like the core issue is still being misunderstood. It's really not a matter of taste. The current implementation of `graph` takes ***quadratic time***, because for every edge in `es` it makes a linear pass over `vs`. So, the running time is O(length(vs)*length(es)).

All I'm proposing is a method that does the same thing and runs in time O(length(vs) + length(es)), that is, ***linear time***. It's really not subtle. 

There are other linear time algorithms; I'm happy to discuss which linear time method is most suitable. However, right now the main way to construct graphs in Graphs.jl takes ***quadratic time*** in the size of its inputs. It shouldn't be down to the user to have to read the source code to realize that the constructor takes quadratic time and manually find a workaround. A fast constructor is exactly what the library should be providing to begin with. By analogy, imagine the shortest path implementation took ***quadratic time*** because it was inspecting every single vertex in the graph and computing the distance from that vertex to the given vertex.

**Comment by mlubin** [comment]
I understand the complexity issues, but the problem isn't just in graph construction. The implementation of dijkstra also uses ``vertex_index``:
```julia
    for e in out_edges(u, graph)
        v::V = target(e, graph)
        iv::Int = vertex_index(v, graph)
        ...
```

**Comment by mrtzh** [comment]
Okay, if you're telling me that the shortest path implementation also takes quadratic time, then I will completely lose faith now ;-)

Let's deal with one issue at a time though. We both understand linear vs quadratic running times and we agree that constructing a graph data structure from a list of vertices and edges should take linear time and not quadratic time. Let's discuss how best to fix it. 

If I understand you correctly, we both think the best way is to augment the general data structure with a hash table. Correct? 

As you point out, this has a problem with vertex types that are not hashable. For such types we could fall back to the linear search over the vertex list.

**Comment by pozorvlak** [comment]
Yeah, a quadratic-by-default constructor is clearly a Bad Thing. There may be users who'd put up with O(n^2) constructors and an extra factor of n in the complexity of every function that calls `vertex_index` because they can't afford the extra memory of a vertex_index hash, but that sounds like an unusual set of requirements which shouldn't be the default. So currently I think the best fix is to add a vertex_index hash to the default graph type. @mrtzh, could you please submit a pull request?

**Comment by mlubin** [comment]
Yes, this seems like a case of Graphs being too general that it's not useful anymore. ``simple_graph`` is fast because the vertex objects are the integer indices themselves, but any other graph that doesn't use ``KeyVertex`` will have slow ``vertex_index``.

**Comment by mrtzh** [comment]
Yes. That's exactly right. Thanks, guys. I will put in the pull request shortly. In the meantime, I'm happy to hear more suggestions on the most elegant way to fix this. The current proposal is a hash table in the general data structure and to have `vertex_index` access that hash table.

**Comment by mlubin** [comment]
Addressed by #161
